# 02 — Data Cleaning

## Objective
Address all data quality issues identified in the exploration phase 
and produce a clean dataset ready for RFM analysis.

## Contents
1. Remove duplicate rows
2. Drop missing CustomerIDs
3. Filter out cancelled invoices
4. Remove zero and negative prices
5. Save cleaned dataset

---

In [1]:
import pandas as pd

df= pd.read_excel('../data/raw/online_retail_II.xlsx', sheet_name = 'Year 2010-2011')

In [2]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

In [3]:
df_clean = df.copy()
df_clean = df_clean.drop_duplicates()
print(f"Duplicates removed: {df.shape[0] - df_clean.shape[0]}")

Duplicates removed: 5268


In [4]:
before_rows = df_clean.shape[0]
df_clean = df_clean.dropna(subset=['Customer ID'])
print(f"Null values removed: {before_rows - df_clean.shape[0]}")

Null values removed: 135037


In [5]:
df_clean.isna().sum()

Invoice        0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
Price          0
Customer ID    0
Country        0
dtype: int64

In [6]:
df_clean.duplicated().sum()

np.int64(0)

In [8]:
df_clean.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='str')

In [9]:
before_rows = df_clean.shape[0]
df_clean = df_clean[df_clean['Invoice'].str.startswith('C').fillna(False)==False]
print(f"Rows with cancelled invoices removed: {before_rows - df_clean.shape[0]}")

Rows with cancelled invoices removed: 8872


In [10]:
before_rows = df_clean.shape[0]
print(f"Zero/negative price rows: {df_clean[df_clean['Price'] <= 0].shape[0]}")
df_clean = df_clean[df_clean['Price'] > 0]
print(f"Rows removed: {before_rows - df_clean.shape[0]}")

Zero/negative price rows: 40
Rows removed: 40


In [13]:
print(f"Dataset shape after cleaning: {df_clean.shape}")

Dataset shape after cleaning: (392693, 8)


## Cleaning Summary

| Step | Issue | Rows Removed |
|------|-------|--------------|
| 1 | Duplicate rows | 5,268 |
| 2 | Missing Customer IDs | 135,037 |
| 3 | Cancelled invoices | 8,872 |
| 4 | Zero/negative prices | 40 |
| **Total** | | **149,217** |

Remaining rows: 392693

In [15]:
df_clean.to_csv('../data/processed/df_clean.csv', index = False)